rfd/0022-ssh-agent-forwarding.md

```diff
@@ -0,0 +1,224 @@
+
+The `tsh ssh` client _also_ offers a `-A` option to forward a Key Agent to the Remote
+Machine, but it will only ever forward the Teleport Key Agent, not the User Key Agent.
+This is surprising behaviour to a user accustomed to OpenSSH. The user finds that keys
+they expected to be available to them in the session on the Remote Machine are not
+available.
+
+The tricky part is that _both_ of these behaviours are sensible, depending on the user's
+expectations. Any solution will have to accommodate both behaviours.
+
+## Details
+
+### Proposed Solution
+
+Behind the scenes, both the OpenSSH client & `tsh ssh` treat the `-A` flag as shorthand
+for setting the SSH config value `ForwardAgent` to `yes`.
+
+Looking at the `man` page for `ssh_config(5)`, we can see the definitive list of values
+that `ForwardAgent` can take, and their meaning:
+
+> **ForwardAgent**
+>
+>   Specifies whether the connection to the authentication agent (if any) will be forwarded
+>   to the remote machine.  The argument may be yes, no (the default), an explicit path to
+>   an agent socket or the name of an environment variable (beginning with ‘$’) in which to
+>   find the path.
+
+The `tsh ssh` client already supports the `yes` and `no` options, where `yes` means
+forwarding the Teleport Key Agent. To allow the user to forward their own Key Agent, _and_
+to bring `tsh ssh`'s behaviour more in line with the OpenSSH client, I propose the following
+values be allowed for the `ForwardAgent`option:
+
+
+| Value          | Interpretation                                          |
+|----------------|---------------------------------------------------------|
+| `no` (default) | `tsh ssh` will not forward _any_ Key Agent              |
+| `yes`          | `tsh ssh` will forward the User Key Agent (if present)  |
+| `local`        | `tsh ssh` will forward the Teleport Key Agent           |
+
+
+* The value `yes` is redefined to refer to the User Key Agent instead of
+  the Teleport Key Agent.
+
+* The value `local` is a `tsh`-specific extension that will activate the
+  backwards-compatible behaviour.
+
+The effect of this change is that the `tsh ssh` option `-A` will automatically acquire
+semantics in line with the OpenSSH client, while the backwards compatible behaviour can be
+activated with something like:
+
+```bash
+$ tsh ssh -o "ForwardAgent local" root@example.com
+```
+
+### Precedence
+
+If both `-A` and `-o "ForwardAgent $VALUE"` are specified together on the same command line, `-A`
+will take precedence. This is consistent with the existing behaviour of `tsh ssh`, where a
+command line like...
+```bash
+$ tsh ssh -A -o "ForwardAgent no" root@example.com
+```
+...would result in the Key Agent being forwarded.
+
+
+### No User Key Agent
+
+If no User Key Agent is running and the user specifies `-A` and/or `-o "ForwardAgent yes"`, then
+`tsh ssh` will NOT forward an Key Agent to the remote machine, consistent with the behaviour of
+OpenSSH.
+
+### Security Concerns
```

Comment:
I think this section can be removed.

Reply:
OK, happy to remove. I think we should document the fact that we thought about the security concerns _somewhere_, though. Perhaps in the PR comment chain?

Reply:
Moved security discussion to PR description